rfc/egress-billing.md

```diff
@@ -0,0 +1,309 @@
+request to the rate limits service and a small amount of computation to decide whether to serve content. It will also give users an
+extremely flexible and extensible mechanism to control how and when their content is served, including support for
+`Origin`-gating and other CDN-esque tools to control their egress costs.
+
+```mermaid
+sequenceDiagram
+  Alice->>Gateway: Do you have bafk...7fi?
+  Gateway->>Rate Limit Service: Has bafk...7fi exceeded rate limits?
+  Rate Limit Service-->>Gateway: YES or NO or UNKNOWN
+  Gateway-)Accounting Service: Serving/not serving bafk...7fi
+  Gateway-->>Alice: 200 if YES or UNKNOWN, 429 if NO
+  Accounting Service->>Rate Limit Service: Update rate limits
+```
+
+## Introduction
+
+We currently operate the IPFS gateway at https://w3s.link as a free service. As a result, it is an unbound source of "cost"
+for our business. Covering some amount of the costs associated with this service make sense as a marketing and community
+building expense, but operating it for free makes it prohibitively expensive to give our customers acceptable read performance
+for their applications through our gateway.
+
+In order to fix this, we propose a design for billing our customers for data egress related to their applications. The suggested
+design allows us to maintain extremely fast read performance by allowing us to configure the "acceptable level" of "unpaid"
+data egress and giving our customers a way to authorize paid data egress using the "content claims" returned by the 
+[blob upload service].
+
+## Rate Limits
+
+Business realities (runway, need for marketing and community growth, etc) will determine an acceptable level of data egress
+through our gateway that can be paid for by the business. We will call this _R_, the "number of requests we're willing to pay for."
+
+We will use a tool like Redis to implement this rate limit, because it has a long history of being used for this and
+the right set of primitives to support it well: https://redis.io/glossary/rate-limiting/ - we will use Redis as an example
+in this document but recommend researching other tools like memcached and Caddy before making a final implementation decision.
+
+When the gateway receives a request for some content, it will make a request to Redis for the relevant rate limit data and decide to
+serve the request based on the response. In order to maintain acceptable read performance, we recommend avoiding any other network
+IO in making the decision to serve a piece of content. In the event that content is served as a result of the rate limit caching service
+being out of date, we count that request against _R_. 
+
+The following TypeScript pseudocode illustrates the rate limiting algorithm:
+
+
+```ts
+enum RateLimited {
+  Yes,
+  No,
+  Maybe,
+}
+
+async function handle(request){
+  const rateLimited = checkRedisToDetermineIfRateLimited(request)
```

Comment:
ah I hear you but I am trying to convey that _only_ "Redis" should consulted here - I've come to think of "Redis" as an API not a product (because that seems to be the reality of it in 2024) - I do mention earlier that I'm using Redis as an example and I think this is consistent with that, so I'd like to keep it